src/GraphQLParser/ParserContext.Parse.cs

```diff
@@ -908,50 +908,28 @@ private GraphQLName ParseName()
 
         private ASTNode? ParseNamedDefinition()
         {
-            var value = _currentToken.Value;
-
-            if (value == "query")
-                return ParseOperationDefinition();
-
-            if (value == "mutation")
-                return ParseOperationDefinition();
-
-            if (value == "subscription")
-                return ParseOperationDefinition();
-
-            if (value == "fragment")
-                return ParseFragmentDefinition();
-
-            if (value == "schema")
-                return ParseSchemaDefinition();
-
-            if (value == "scalar")
-                return ParseScalarTypeDefinition();
-
-            if (value == "type")
-                return ParseObjectTypeDefinition();
-
-            if (value == "interface")
-                return ParseInterfaceTypeDefinition();
-
-            if (value == "union")
-                return ParseUnionTypeDefinition();
-
-            if (value == "enum")
-                return ParseEnumTypeDefinition();
-
-            if (value == "input")
-                return ParseInputObjectTypeDefinition();
-
-            if (value == "extend")
-                return ParseTypeExtension();
-
-            if (value == "directive")
-                return ParseDirectiveDefinition();
+            return ExpectOneOf(TopLevelKeywordOneOf, advance: false) switch
+            {
+                "query" => ParseOperationDefinition(),
+                "mutation" => ParseOperationDefinition(),
+                "subscription" => ParseOperationDefinition(),
+                "fragment" => ParseFragmentDefinition(),
+                "schema" => ParseSchemaDefinition(),
+                "scalar" => ParseScalarTypeDefinition(),
+                "type" => ParseObjectTypeDefinition(),
+                "interface" => ParseInterfaceTypeDefinition(),
+                "union" => ParseUnionTypeDefinition(),
+                "enum" => ParseEnumTypeDefinition(),
+                "input" => ParseInputObjectTypeDefinition(),
+                "extend" => ParseTypeExtension(),
+                "directive" => ParseDirectiveDefinition(),
 
-            return null;
+                _ => throw new NotSupportedException("Compiler never gets here since ExpectOneOf throws.")
+            };
         }
 
+        // TODO: 1. May be changed to use ExpectOneOf, or
```

Comment:
@Shane32  I can modify `ExpectOneOf` to get `token`. Then it will be possible to call `ExpectOneOf` from `ParseNamedDefinitionWithDescription`. What do you think?

Reply:
You decide; I think you understand the needs here better than me.

My goals for GraphQL.NET v5 are changes to AutoRegistering types and integrating deserialization to `IDocumentWriter`, which I think will bring great benefits to users.

If you get a chance to review my notes and just a couple files I mentioned in https://github.com/graphql-dotnet/graphql-dotnet/pull/2831, I'd like your input so I can continue working on the PR.  See https://github.com/graphql-dotnet/graphql-dotnet/pull/2831#issuecomment-1004441109

Reply:
They're big. I postponed them. First I want to finish things with a parser and small things.

Reply:
Please do not hold it up long.  I cannot continue working until you take the time to briefly review the 6 files I mentioned.  It is not much to review.

(Once the PR is ready for a full review I will let you know.)

Reply:
> First I want to finish things with a parser and small things.

👍

Reply:
![изображение](https://user-images.githubusercontent.com/21261007/148124190-5a2fda9a-a085-437b-8d4b-5786772d0252.png)

Reply:
yup.  that's why I only asked for a review of 6 files

Reply:
![image](https://user-images.githubusercontent.com/6377684/148124640-10b9aa75-3495-473b-971d-4edb9961da0d.png)

Reply:
I have 4 more weekends ahead.